src/pool/srv_pool.c

```diff
@@ -7438,6 +7439,117 @@ ds_pool_tgt_revert_rebuild(uuid_t pool_uuid, struct pool_target_id_list *list)
 					NULL, NULL, NULL, NULL, NULL, NULL);
 }
 
+struct safe_reint_ult_arg {
+	struct ds_pool *pool;
+	ABT_eventual    eventual;
+};
+
+static void
+ds_rebuild_abort_ult(void *arg)
+{
+	struct safe_reint_ult_arg *ult_arg = arg;
+	int                        rc      = 0;
+	uint32_t                   version = 0;
+
+	ds_rebuild_running_query(ult_arg->pool->sp_uuid, RB_OP_REBUILD, &version, NULL, NULL);
+	if (version == 0) {
+		D_ERROR(DF_UUID
+			": No rebuild process is running, --no-migration option is not permitted",
+			DP_UUID(ult_arg->pool->sp_uuid));
+		rc = -DER_NO_PERM;
+		goto out;
+	}
+
+	ds_rebuild_abort(ult_arg->pool->sp_uuid, -1, -1, -1);
+
+out:
+	ABT_eventual_set(ult_arg->eventual, (void *)&rc, sizeof(rc));
+}
+
+static int
+ds_pool_tgt_safe_reint(struct pool_svc *svc, struct pool_target_addr_list *list,
+		       struct pool_target_addr_list *inval_list_out, uint32_t *map_version,
+		       struct rsvc_hint *hint, enum map_update_source src, uint32_t flags)
+{
+	int                        rc;
+	struct safe_reint_ult_arg  arg;
+	ABT_eventual               eventual;
+	int                       *status;
+	bool                       updated;
+	struct pool_target_id_list target_list = {0};
+	uint32_t                   tgt_map_ver = 0;
+	daos_epoch_t               rebuild_eph = d_hlc_get();
+	bool no_data_sync = (svc->ps_pool->sp_reint_mode == DAOS_REINT_MODE_NO_DATA_SYNC);
+
+	if (!no_data_sync) {
+		d_rank_list_t *dead_rank_list = NULL;
+		/*
+		 * Safe reintegration mode is only allowed when:
+		 * 1. Pool rebuild is in progress
```

Comment:
I'm trying to remember the exact scenario. There may be a case where pool rebuild is *not* running, but the admin will want to perform a reintegrate --no-migration to recover the pool.

let's say DAOS_POOL_RF=2, and we have failure event A with 2 fault domains affected - 2 nodes engines being excluded from the pool. This goes for some time, and admin decides to shut down the system before the rebulid finishes. Then, on system restart, a third node has an immediate failure (failure event B). I think in this condition, the pool_map_crit_prompt() function will emit the urgent message alerting the administrator that a rebuild will not be started for this latest pool map change. So, some admin command will be needed to cause a rebuild to be triggered.

Reply:
After restarting, system will resume event A's rebuild? or rebuild will not be triggered because of event B and it broke RF.

Reply:
I think it will resume event A rebuild after restarting. However, IIRC, the scenario was that these rebuilds could fail (-DER_NOMEM failures occurred on rebuilds of certain pools). If this happens before dmg pool reintegrate --no-migration is run, then I guess the rebuild would not be triggered. Maybe this would require a restart and then having the admin run the reintegrate --no-migration while the event A excludes rebuild is running?

Maybe we can come up with a test exercise to see how this would work out.